spartan_parallel/src/sparse_mlpoly.rs

```diff
@@ -1,98 +1,89 @@
 #![allow(clippy::type_complexity)]
 #![allow(clippy::too_many_arguments)]
 #![allow(clippy::needless_range_loop)]
+use crate::scalar::SpartanExtensionField;
+
 use super::dense_mlpoly::DensePolynomial;
-use super::dense_mlpoly::{
-  EqPolynomial, IdentityPolynomial, PolyCommitment, PolyCommitmentGens, PolyEvalProof,
-};
+use super::dense_mlpoly::{EqPolynomial, IdentityPolynomial, PolyEvalProof};
 use super::errors::ProofVerifyError;
 use super::math::Math;
 use super::product_tree::{DotProductCircuit, ProductCircuit, ProductCircuitEvalProofBatched};
 use super::random::RandomTape;
-use super::scalar::Scalar;
 use super::timer::Timer;
 use super::transcript::{AppendToTranscript, ProofTranscript};
 use core::cmp::Ordering;
 use merlin::Transcript;
 use serde::{Deserialize, Serialize};
 
 #[derive(Clone, Debug, Serialize, Deserialize)]
-pub struct SparseMatEntry {
+pub struct SparseMatEntry<S: SpartanExtensionField> {
   row: usize,
   col: usize,
-  val: Scalar,
+  val: S,
```

Comment:
If `S` itself extends the `ExtensionField` trait, then val's type should be `S::BaseField`. Otherwise, we need to use this cumbersome expression `<S::InnerType as ExtensionField>::BaseField`.